### PR TITLE
Optimize heartbeat - async process statistics

### PR DESCRIPTION
**Pull request by nolouch**
### What problem does this PR solve?

Issue Number: Ref https://github.com/tikv/pd/issues/7897

### What is changed and how does it work?

```commit-message
Optimize heartbeat process
- Introduce `SyncRunner` and `AsyncRunner`
- Async the statics updates on subtree
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)

Before:
image
image

After:
image
image

The duration of the pending state on tikvs has been reduced from 10 minutes to 3 minutes

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/ratelimit/runner.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+		}
+		f(ctx)
+	}(ctx, opt)
+
+	return nil
+}
+
+func (s *AsyncRunner) recover() {
+	if r := recover(); r != nil {
+		log.Error("panic in runner", zap.Any("error", r))
```
Comment (HuSharp) [inline pkg/ratelimit/runner.go:93]: Should we add more information? such as task name.

File: pkg/core/region.go [hunk 2]
```diff
@@ -933,48 +967,102 @@ func (r *RegionsInfo) PutRegion(region *RegionInfo) []*RegionInfo {
...
+	tracer.OnSetRegionFinished()
 	r.UpdateSubTree(region, origin, overlaps, rangeChanged)
-	trace.OnUpdateSubTreeFinished()
+	tracer.OnUpdateSubTreeFinished()
 	return overlaps, nil
 }
 
+// CheckAndPutSupTree checks if the region is valid to put to the root, if valid then return error.
+// Usually used with CheckAndPutSubTree together.
+func (r *RegionsInfo) CheckAndPutSupTree(ctx context.Context, region *RegionInfo) ([]*RegionInfo, error) {
```
Comment (rleungx) [inline pkg/core/region.go:1014]:
```suggestion
func (r *RegionsInfo) CheckAndPutSubTree(ctx context.Context, region *RegionInfo) ([]*RegionInfo, error) {
```

File: pkg/ratelimit/concurrency_limiter.go [hunk 3]
```diff
@@ -54,28 +62,28 @@ func (l *concurrencyLimiter) release() {
...
 	defer l.mu.Unlock()
 
 	l.limit = limit
 }
 
-func (l *concurrencyLimiter) getCurrent() uint64 {
-	l.mu.RLock()
-	defer l.mu.RUnlock()
+func (l *ConcurrencyLimiter) getCurrent() uint64 {
+	l.mu.Lock()
```
Comment (rleungx) [inline pkg/ratelimit/concurrency_limiter.go:80]: why change it?

File: pkg/ratelimit/concurrency_limiter.go [hunk 4]
```diff
@@ -84,3 +92,65 @@ func (l *concurrencyLimiter) getMaxConcurrency() uint64 {
...
 }
+
+// GetRunningTasksNum returns the number of running tasks.
+func (l *ConcurrencyLimiter) GetRunningTasksNum() uint64 {
+	return l.getCurrent()
+}
+
+// GetWaitingTasksNum returns the number of waiting tasks.
+func (l *ConcurrencyLimiter) GetWaitingTasksNum() uint64 {
+	l.mu.Lock()
```
Comment (rleungx) [inline pkg/ratelimit/concurrency_limiter.go:103]: rlock?

File: pkg/ratelimit/runner.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
+// Copyright 2022 TiKV Project Authors.
```
Comment (rleungx) [inline pkg/ratelimit/runner.go:1]:
```suggestion
// Copyright 2024 TiKV Project Authors.
```

File: pkg/ratelimit/runner.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+	Limit    *ConcurrencyLimiter
+}
+
+// RunTask except the callback f is run in a goroutine.
+// The call doesn't block for the callback to finish execution.
+func (s *AsyncRunner) RunTask(ctx context.Context, opt TaskOpts, f func(context.Context)) error {
+	if opt.Limit != nil && atomic.LoadInt64(&s.numTasks) >= int64(s.maxPendingTasks) {
+		return ErrMaxWaitingTasksExceeded
+	}
+	s.addTask(1)
```
Comment (rleungx) [inline pkg/ratelimit/runner.go:59]: Will it keep adding without executing the defer?

**Comment by nolouch** [comment]
will split this pr, close.

**Pull request closed without merging**